### PR TITLE
docs: add High Availability information for s3gw and Longhorn

### PR DESCRIPTION
**Pull request by giubacc**
add High Availability information for s3gw and Longhorn

Fixes: https://github.com/aquarist-labs/s3gw/issues/841

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] CHANGELOG.md has been updated should there be relevant changes in this PR.

### CODE REVIEW

File: docs/high-availability.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+The s3gw mainly relies on Kubernetes and Longhorn to achieve High Availability (HA).
+
+A single s3gw instance is attached to its private Longhorn-provided persistent
+volume (PV) with Kubernetes restarting/redeploying as necessary on failure.
+
+In the context of the s3gw and Longhorn, this HA model is called: *Active/Standby*.
+
+The PV's state is guaranteed by Longhorn ensuring that a PV, along with
+its content, to be always available on any cluster's node.
```
Comment (tserong) [inline docs/high-availability.md]:
```suggestion
its content, will always be available on all cluster nodes.
```

File: docs/high-availability.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+The PV's state is guaranteed by Longhorn ensuring that a PV, along with
+its content, will always be available on all cluster nodes.
+This means that the s3gw can mount its Longhorn PV regardless of the
+current scheduling on the cluster.
+
+Obviously, the s3gw can't achieve higher availability or reliability than
+the underlying Longhorn volumes.
+This means that any corruption should occur in the file system is outside what
+s3gw can reasonably protect against; that's all undefined behavior and "restore
+from backup" time.
```
Comment (jecluis) [inline docs/high-availability.md:19]: I think we can drop the `that's all undefined behavior ...`, or maybe replace it with just `In this case, restoring from backup might be the only option.`.

File: docs/high-availability.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+The *Active/Standby* model claims the following characteristics:
```
Comment (jecluis) [inline docs/high-availability.md:21]: s/claims/offers/

File: docs/high-availability.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+  the backend pod.
+
+- s3gw pod rescheduling: this case applies when Kubernetes decides to reschedule
+  the pod to another node, eg: when the node goes low on resources. This would often
+  be called a "switch-over" in an HA cluster - e.g., an administratively orchestrated
+  transition of the service to a new node (say, for maintenance/upgrade/etc reasons).
+  This has the advantage of being schedulable, so it can happen at times of low load
+  if these exist.
+
+When any of these scenarios should happen, Kubernetes restarts the s3gw pod and we
```
Comment (jecluis) [inline docs/high-availability.md:41]: s/should happen/happens/

File: docs/high-availability.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+expect the service to be restored in seconds.
+
+## Non-graceful node failure
+
+Currently, Kubernetes ([1.28](https://kubernetes.io/releases/) at the time of
+writing this), does not automatically restart a pod attached to a RWO volume
+in the event that the node running it suffers a failure.
```
Comment (jecluis) [inline docs/high-availability.md:48]: I'd write "in case the node it is running on suffers from a failure"

File: docs/high-availability.md [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+Reasons behind this behavior is that workloads, such as RWO volumes require
```
Comment (jecluis) [inline docs/high-availability.md:49]: add a comma between "RWO volumes" and "require"?

File: docs/high-availability.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+*at-most-one* semantics.
+Failures affecting these kind of workloads risk data loss and/or corruption
+if nodes (and the workloads running on them) are wrongly assumed to be dead.
+For this reason it is crucial to know that the node has reached a safe state
+before initiating recovery of the workload.
```
Comment (jecluis) [inline docs/high-availability.md:54]: s/recovery of the workload/workload recovery/

File: docs/high-availability.md [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+Longhorn offers the option to perform a [Pod Deletion Policy][pod-deletion-policy]
+when a node should go down unexpectedly.
```
Comment (jecluis) [inline docs/high-availability.md:57]: I'd write instead

```
offers the option to define a Pod Deletion Policy when the node goes down unexpectedly
```

File: docs/high-availability.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+This means that Longhorn will force delete StatefulSet/Deployment terminating pods
```
Comment (jecluis) [inline docs/high-availability.md:58]: you mean "force-delete" or something? or maybe "forcefully delete"?

File: docs/high-availability.md [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+on nodes that are down to release Longhorn volumes so that Kubernetes
+can spin up replacement pods.
+
+Anyway, when employing this mitigation, the user must be aware that assuming a node
```
Comment (jecluis) [inline docs/high-availability.md:62]: Maybe drop the "Anyway" and start with "When employing ..." ?

File: docs/high-availability.md [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+dead could lead to data loss and/or corruption if that assumption was actually untrue.
+
+The s3gw and the Longhorn team is currently investigating some
+[hypotheses of solutions][longhorn-issue-1]
+to address this problem at its roots.
```
Comment (jecluis) [inline docs/high-availability.md:67]: s/roots/root/

**Review by jecluis — CHANGES_REQUESTED** [review]

**Pull request closed without merging**